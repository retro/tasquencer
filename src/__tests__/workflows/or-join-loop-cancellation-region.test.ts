import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../../index.js';
import { getEnabledTaskNames, makeIdGenerator } from '../shared.js';

it('runs a net with an "or" join, a loop and a cancellation region', ({
  expect,
}) => {
  const workflowDefinition = Builder.workflow()
    .withName('or-join-cancellation-region')
    .startCondition('start')
    .task('A', Builder.emptyTask())
    .task('B', Builder.emptyTask())
    .task('C', Builder.emptyTask())
    .task('D', Builder.emptyTask().withSplitType('and'))
    .task('E', Builder.emptyTask().withJoinType('or'))
    .condition('c1')
    .condition('c2')
    .condition('c3')
    .endCondition('end')
    .connectCondition('start', (to) => to.task('A'))
    .connectTask('A', (to) => to.condition('c1'))
    .connectCondition('c1', (to) => to.task('B'))
    .connectTask('B', (to) => to.condition('c2'))
    .connectCondition('c2', (to) => to.task('C').task('E'))
    .connectTask('C', (to) => to.condition('c3'))
    .connectCondition('c3', (to) => to.task('D').task('E'))
    .connectTask('D', (to) => to.condition('c1').condition('c2'))
    .connectTask('E', (to) => to.condition('end'))
    .cancellationRegion('C', { tasks: ['B'], conditions: ['c1', 'c2'] });

  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    expect(yield* service.getState()).toMatchSnapshot();

    yield* service.startRootWorkflow();
    const state1 = yield* service.getState();
    expect(state1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1)).toEqual(new Set(['A']));

    yield* service.startTask('A', {});
    const state2 = yield* service.getState();
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(new Set(['B']));

    yield* service.startTask('B', {});
    const state3 = yield* service.getState();
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(new Set(['C', 'E']));

    yield* service.startTask('C', {});
    const state4 = yield* service.getState();
    expect(state4).toMatchSnapshot();
    expect(getEnabledTaskNames(state4)).toEqual(new Set(['D']));

    yield* service.startTask('D', {});
    const state5 = yield* service.getState();
    expect(state5).toMatchSnapshot();
    expect(getEnabledTaskNames(state5)).toEqual(new Set(['B', 'C', 'E']));

    yield* service.startTask('E', {});
    const state6 = yield* service.getState();
    expect(state6).toMatchSnapshot();
    expect(getEnabledTaskNames(state6)).toEqual(new Set());
    expect(state6.workflows[0]?.state).toBe('completed');
  });

  Effect.runSync(program);
});
