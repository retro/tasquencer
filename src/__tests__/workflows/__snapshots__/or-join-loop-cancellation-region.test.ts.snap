// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runs a net with an "or" join, a loop and a cancellation region 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "A",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "B",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "C",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "D",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "initialized",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 2`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "A",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "B",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "C",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "D",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "A",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "B",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "C",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "D",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "A",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "B",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "C",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "D",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 5`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "A",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "B",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "C",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "D",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 6`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "A",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "B",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "C",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "D",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 0,
      "name": "E",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with an "or" join, a loop and a cancellation region 7`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "c1",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c2",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 0,
      "name": "c3",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "A",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "B",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "C",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "D",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
    {
      "generation": 1,
      "name": "E",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "or-join-cancellation-region",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "or-join-cancellation-region",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;
